provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,211 @@
+//
+// For example, if there is a property "someProperty" that is both an input and
+// an output, the underlying type could have readonly properties meaning that part
+// of the type is an output only value. Those will not exist as input values
+// and will be marked as computed
+func previewResourceOutputs(
+	inputs resource.PropertyMap,
+	types map[string]metadata.CloudAPIType,
+	outputs map[string]schema.PropertySpec,
+	readOnly []string,
+) resource.PropertyMap {
+	result := resource.PropertyMap{}
+	for name, prop := range outputs {
+		key := resource.PropertyKey(name)
+		if inputValue, ok := inputs[key]; ok {
+			result[key] = previewOutputValue(inputValue, types, &prop.TypeSpec, filterAndReturnNested(readOnly, name))
+			// otherwise we could have one of two cases
+			// 1. the property is an input & output property, but does not have an input value
+			// 2. the property is a readonly (output only) property
+			// In either case, the property could be computed so we should default to marking it as computed
+		} else if isReadOnly(readOnly, name) {
+			result[key] = resource.MakeComputed(resource.NewStringProperty(""))
+		}
+	}
+	return result
+}
+
+// populateStableOutputs updates the preview outputs with the outputs from
+// the prior state, but only in certain cases.
+//
+// There is no way (based on the schema) to tell which output properties are stable
+// (i.e. they will only change if the resource is replaced). Because of this we
+// cannot blanket copy all outputs from the prior state. The ones that are not stable
+// should remain computed since they could be changing.
+//
+// Instead, we use some heuristics to determine which outputs are stable and which are not.
+// We are making the assumption that Name, Id, and Arn are stable outputs. Sometimes resources
+// will have `name`, `id`, and `arn` properties, and sometimes they will have `resourceName`, `resourceId`,
+// and `resourceArn` properties.
+func populateStableOutputs(
+	outputsFromInputs resource.PropertyMap,
+	outputsFromPriorState resource.PropertyMap,
+	readOnly []string,
+	resourceTypeName tokens.TypeName,
+) resource.PropertyMap {
+	for _, readOnlyProp := range readOnly {
+		// For now skip outputs that are part of an array property.
+		// We have no way of knowing which item in the outputs array corresponds
+		// to which item from the inputs array (since inputs could be changing).
+		if strings.Contains(readOnlyProp, "/*/") {
+			continue
+		}
+		if isStableOutput(readOnlyProp, resourceTypeName) {
+			// nested property. Sometimes the resource Arn is in a nested property
+			if strings.Contains(readOnlyProp, "/") {
+				props := strings.Split(readOnlyProp, "/")
+				current := props[0]
+				key := resource.PropertyKey(current)
+				if outputFromInput, ok := outputsFromInputs[key]; ok {
+					if outputFromState, ok := outputsFromPriorState[key]; ok {
+						outputsFromInputs[key] = resource.NewObjectProperty(populateStableOutputs(
```

Comment:
Will this only fire up when both inputs and prior state is present? Curious if this will exclude the case when an input is present but output not.. Ah ok that may be expected because this is only working with 'stable' info propagation. 

How about when a state entry is present but the input entry is not? Why will it skip this?

Reply:
I think I've changed my mind about a previous comment you made about this being part of `previewResourceOutputs`. This function is meant to only be called with the output of `previewResourceOutputs` so `outputsFromInputs` contains all the possible output properties already. 

If an entry is not present in the input, but is in the prior state then it will be marked as computed (as part of the `previewResourceOutputs` function.